flink-connectors/flink-connector-elasticsearch7/src/main/java/org/apache/flink/connector/elasticsearch/sink/ElasticsearchWriter.java

```diff
@@ -0,0 +1,326 @@
+                                networkClientConfig));
+        this.bulkProcessor =
+                configureBulkProcessor(
+                        BulkProcessor.builder(
+                                (bulkRequest, bulkResponseActionListener) ->
+                                        client.bulkAsync(
+                                                bulkRequest,
+                                                RequestOptions.DEFAULT,
+                                                bulkResponseActionListener),
+                                new BulkListener()),
+                        bulkProcessorConfig);
+        this.requestIndexer = new DefaultRequestIndexer();
+        checkNotNull(metricGroup);
+        metricGroup.setCurrentSendTimeGauge(() -> receiveTime - lastSendTime);
+        this.numBytesOutCounter = metricGroup.getIOMetricGroup().getNumBytesOutCounter();
+        try {
+            processor.open();
+        } catch (Exception e) {
+            throw new RuntimeException("Failed to open the ElasticsearchProcessor", e);
+        }
+    }
+
+    @Override
+    public void write(IN element, Context context) throws IOException, InterruptedException {
+        // do not allow new bulk writes until all actions are flushed
+        while (checkpointInProgress) {
+            mailboxExecutor.yield();
+        }
+        processor.process(element, context, requestIndexer);
+    }
+
+    @Override
+    public List<Void> prepareCommit(boolean flush) throws IOException, InterruptedException {
+        checkpointInProgress = true;
+        while (pendingActions != 0) {
+            bulkProcessor.flush();
+            LOG.info("Waiting for the response of {} pending actions.", pendingActions);
+            mailboxExecutor.yield();
+        }
+        checkpointInProgress = false;
+        return Collections.emptyList();
+    }
+
+    @Override
+    public void close() throws Exception {
+        closed = true;
+        processor.close();
+        client.close();
+    }
+
+    private static RestClientBuilder configureRestClientBuilder(
+            RestClientBuilder builder, NetworkClientConfig networkClientConfig) {
+        if (networkClientConfig.getConnectionPathPrefix() != null) {
+            builder.setPathPrefix(networkClientConfig.getConnectionPathPrefix());
+        }
+        if (networkClientConfig.getPassword() != null
+                && networkClientConfig.getUsername() != null) {
+            final CredentialsProvider credentialsProvider = new BasicCredentialsProvider();
+            credentialsProvider.setCredentials(
+                    AuthScope.ANY,
+                    new UsernamePasswordCredentials(
+                            networkClientConfig.getUsername(), networkClientConfig.getPassword()));
+            builder.setHttpClientConfigCallback(
+                    httpClientBuilder ->
+                            httpClientBuilder.setDefaultCredentialsProvider(credentialsProvider));
+        }
```

Comment:
I think we should allow arbitrary `CredentialsProvider` here. We have made the mistake too often to limit the user.
Let's use a `SerializableSupplier` and expose that to the builder. We can use a subbuilder for basic creds.

Reply:
We have to find a balance here all additional UDFs are making the usage in Table API more difficult. The previous Table API connector only supported Basic auth and no one ever complained.

In general, I understand your idea but we can easily extend the authentication feature set if needed.